File: src/security/patterns/ruby.ts
import { SecurityPattern, VulnerabilityType } from '../types.js';

/**
 * Ruby-specific security patterns for vulnerability detection
 */
export const rubySecurityPatterns: SecurityPattern[] = [
  // SQL Injection patterns
  {
    id: 'ruby-sql-injection-string-interpolation',
    type: VulnerabilityType.SQL_INJECTION,
    pattern: /\.(find_by_sql|execute|exec_query)\s*\(\s*["'`].*#\{.*\}/g,
    severity: 'high',
    message: 'SQL injection vulnerability: String interpolation in SQL queries',
    remediation: 'Use parameterized queries with ? placeholders or ActiveRecord methods',
    cwe: 'CWE-89',
    owasp: 'A03',
    language: 'ruby'
  },
  {
    id: 'ruby-sql-injection-where-string',
    type: VulnerabilityType.SQL_INJECTION,
    pattern: /\.where\s*\(\s*["'`].*#\{.*\}/g,
    severity: 'high',
    message: 'SQL injection vulnerability: String interpolation in where clause',
    remediation: 'Use parameterized where clauses: where("name = ?", name)',
    cwe: 'CWE-89',
    owasp: 'A03',
    language: 'ruby'
  },

  // Command Injection patterns
  {
    id: 'ruby-command-injection-system',
    type: VulnerabilityType.COMMAND_INJECTION,
    pattern: /system\s*\(\s*["'`].*#\{.*\}/g,
    severity: 'critical',
    message: 'Command injection vulnerability: String interpolation in system()',
    remediation: 'Use system() with array of arguments or validate input',
    cwe: 'CWE-78',
    owasp: 'A03',
    language: 'ruby'
  },
  {
    id: 'ruby-command-injection-backticks',
    type: VulnerabilityType.COMMAND_INJECTION,
    pattern: /`.*#\{.*\}`/g,
    severity: 'critical',
    message: 'Command injection vulnerability: String interpolation in backticks',
    remediation: 'Use Open3.capture3() or system() with array arguments',
    cwe: 'CWE-78',
    owasp: 'A03',
    language: 'ruby'
  },
  {
    id: 'ruby-command-injection-exec',
    type: VulnerabilityType.COMMAND_INJECTION,
    pattern: /exec\s*\(\s*["'`].*#\{.*\}/g,
    severity: 'critical',
    message: 'Command injection vulnerability: String interpolation in exec()',
    remediation: 'Use exec() with array of arguments or validate input',
    cwe: 'CWE-78',
    owasp: 'A03',
    language: 'ruby'
  },

  // Deserialization vulnerabilities
  {
    id: 'ruby-unsafe-marshal-load',
    type: VulnerabilityType.INSECURE_DESERIALIZATION,
    pattern: /Marshal\.load/g,
    severity: 'critical',
    message: 'Insecure deserialization: Marshal.load can execute arbitrary code',
    remediation: 'Use JSON.parse or implement custom deserialization with validation',
    cwe: 'CWE-502',
    owasp: 'A08',
    language: 'ruby'
  },
  {
    id: 'ruby-unsafe-yaml-load',
    type: VulnerabilityType.INSECURE_DESERIALIZATION,
    pattern: /YAML\.load(?!_file|_stream)/g,
    severity: 'high',
    message: 'Insecure deserialization: YAML.load can execute arbitrary code',
    remediation: 'Use YAML.safe_load for untrusted input',
    cwe: 'CWE-502',
    owasp: 'A08',
    language: 'ruby'
  },
  {
    id: 'ruby-unsafe-eval',
    type: VulnerabilityType.INSECURE_DESERIALIZATION,
    pattern: /\beval\s*\(/g,
    severity: 'critical',
    message: 'Code injection vulnerability: eval() can execute arbitrary code',
    remediation: 'Avoid eval() or use safe evaluation libraries',
    cwe: 'CWE-95',
    owasp: 'A03',
    language: 'ruby'
  },

  // Path Traversal
  {
    id: 'ruby-path-traversal-file-read',
    type: VulnerabilityType.PATH_TRAVERSAL,
    pattern: /File\.(read|open)\s*\(\s*.*#\{.*\}/g,
    severity: 'medium',
    message: 'Path traversal vulnerability: Unsanitized file paths in File operations',
    remediation: 'Validate and sanitize file paths, use File.join() safely',
    cwe: 'CWE-22',
    owasp: 'A01',
    language: 'ruby'
  },

  // Rails-specific vulnerabilities
  {
    id: 'ruby-rails-mass-assignment',
    type: VulnerabilityType.MASS_ASSIGNMENT,
    pattern: /\.(create|update|update_attributes)\s*\(\s*params\[/g,
    severity: 'high',
    message: 'Mass assignment vulnerability: Unfiltered params in model operations',
    remediation: 'Use strong parameters with permit() to whitelist attributes',
    cwe: 'CWE-915',
    owasp: 'A04',
    language: 'ruby'
  },
  {
    id: 'ruby-rails-redirect-to-params',
    type: VulnerabilityType.OPEN_REDIRECT,
    pattern: /redirect_to\s+params\[/g,
    severity: 'medium',
    message: 'Open redirect vulnerability: Unvalidated redirect_to with params',
    remediation: 'Validate redirect URLs against whitelist of allowed domains',
    cwe: 'CWE-601',
    owasp: 'A01',
    language: 'ruby'
  },

  // XSS in ERB templates
  {
    id: 'ruby-erb-raw-output',
    type: VulnerabilityType.XSS,
    pattern: /<%=\s*raw\s+.*%>/g,
    severity: 'medium',
    message: 'XSS vulnerability: Using raw() in ERB templates',
    remediation: 'Use html_safe only for trusted content, or sanitize with sanitize()',
    cwe: 'CWE-79',
    owasp: 'A03',
    language: 'ruby'
  },
  {
    id: 'ruby-erb-html-safe',
    type: VulnerabilityType.XSS,
    pattern: /<%=.*\.html_safe\s*%>/g,
    severity: 'medium',
    message: 'Potential XSS vulnerability: html_safe in ERB template',
    remediation: 'Ensure content is properly escaped before using html_safe',
    cwe: 'CWE-79',
    owasp: 'A03',
    language: 'ruby'
  },

  // Weak cryptography
  {
    id: 'ruby-weak-hash-md5',
    type: VulnerabilityType.WEAK_CRYPTOGRAPHY,
    pattern: /Digest::MD5/g,
    severity: 'medium',
    message: 'Weak cryptography: MD5 is cryptographically broken',
    remediation: 'Use Digest::SHA256 or Digest::SHA3 for cryptographic hashing',
    cwe: 'CWE-327',
    owasp: 'A02',
    language: 'ruby'
  },
  {
    id: 'ruby-weak-hash-sha1',
    type: VulnerabilityType.WEAK_CRYPTOGRAPHY,
    pattern: /Digest::SHA1/g,
    severity: 'medium',
    message: 'Weak cryptography: SHA-1 is deprecated for security purposes',
    remediation: 'Use Digest::SHA256 or Digest::SHA3 for cryptographic hashing',
    cwe: 'CWE-327',
    owasp: 'A02',
    language: 'ruby'
  },

  // Hardcoded secrets
  {
    id: 'ruby-hardcoded-secret-key',
    type: VulnerabilityType.HARDCODED_SECRETS,
    pattern: /secret_key_base\s*=\s*["'`][a-f0-9]{64,}/g,
    severity: 'critical',
    message: 'Hardcoded secret: Rails secret_key_base in source code',
    remediation: 'Use environment variables or Rails credentials for secrets',
    cwe: 'CWE-798',
    owasp: 'A07',
    language: 'ruby'
  }
];

File: lib/rsolv_api/security/patterns/ruby.ex
defmodule RsolvApi.Security.Patterns.Ruby do
  @moduledoc """
  Ruby security patterns for detecting vulnerabilities.
  
  This module contains 20 security patterns specifically designed for Ruby
  code. Each pattern includes detection rules, test cases, and educational
  documentation.
  """
  
  alias RsolvApi.Security.Pattern
  alias RsolvApi.Security.Patterns.Ruby.MissingAuthentication
  alias RsolvApi.Security.Patterns.Ruby.MassAssignment
  alias RsolvApi.Security.Patterns.Ruby.HardcodedSecrets
  alias RsolvApi.Security.Patterns.Ruby.SqlInjectionInterpolation
  alias RsolvApi.Security.Patterns.Ruby.CommandInjection
  alias RsolvApi.Security.Patterns.Ruby.XpathInjection
  alias RsolvApi.Security.Patterns.Ruby.LdapInjection
  alias RsolvApi.Security.Patterns.Ruby.WeakRandom
  alias RsolvApi.Security.Patterns.Ruby.DebugModeEnabled
  alias RsolvApi.Security.Patterns.Ruby.EvalUsage
  alias RsolvApi.Security.Patterns.Ruby.WeakPasswordStorage
  alias RsolvApi.Security.Patterns.Ruby.UnsafeDeserializationMarshal
  alias RsolvApi.Security.Patterns.Ruby.UnsafeYaml
  alias RsolvApi.Security.Patterns.Ruby.InsufficientLogging
  alias RsolvApi.Security.Patterns.Ruby.SsrfOpenUri
  alias RsolvApi.Security.Patterns.Ruby.XssErbRaw
  alias RsolvApi.Security.Patterns.Ruby.PathTraversal
  alias RsolvApi.Security.Patterns.Ruby.OpenRedirect
  alias RsolvApi.Security.Patterns.Ruby.InsecureCookie
  
  @doc """
  Returns all Ruby security patterns.
  
  ## Examples
  
      iex> patterns = RsolvApi.Security.Patterns.Ruby.all()
      iex> length(patterns)
      20
      iex> Enum.all?(patterns, &match?(%RsolvApi.Security.Pattern{}, &1))
      true
  """
  def all do
    [
      missing_authentication(),
      mass_assignment(),
      weak_crypto_md5(),
      hardcoded_secrets(),
      sql_injection_interpolation(),
      command_injection(),
      xpath_injection(),
      ldap_injection(),
      weak_random(),
      debug_mode_enabled(),
      eval_usage(),
      weak_password_storage(),
      unsafe_deserialization_marshal(),
      unsafe_yaml(),
      insufficient_logging(),
      ssrf_open_uri(),
      xss_erb_raw(),
      path_traversal(),
      open_redirect(),
      insecure_cookie()
    ]
  end
  
  @doc """
  Missing Authentication in Rails Controller pattern.
  
  Detects Rails controllers without authentication filters.
  
  ## Examples
  
      iex> pattern = RsolvApi.Security.Patterns.Ruby.missing_authentication()
      iex> pattern.id
      "ruby-broken-access-control-missing-auth"
      iex> pattern.severity
      :high
  """
  defdelegate missing_authentication(), to: MissingAuthentication, as: :pattern
  
  @doc """
  Mass Assignment Vulnerability pattern.
  
  Detects unfiltered params in model operations.
  
  ## Examples
  
      iex> pattern = RsolvApi.Security.Patterns.Ruby.mass_assignment()
      iex> pattern.type
      :mass_assignment
  """
  defdelegate mass_assignment(), to: MassAssignment, as: :pattern
  
  @doc """
  Hardcoded Secrets pattern.
  
  Detects hardcoded API keys, passwords, and secrets.
  
  ## Examples
  
      iex> pattern = RsolvApi.Security.Patterns.Ruby.hardcoded_secrets()
      iex> pattern.severity
      :critical
  """
  defdelegate hardcoded_secrets(), to: HardcodedSecrets, as: :pattern
  
  @doc """
  Weak Cryptography - MD5 Usage pattern.
  
  Detects usage of weak MD5 hash algorithm.
  
  ## Examples
  
      iex> pattern = RsolvApi.Security.Patterns.Ruby.weak_crypto_md5()
      iex> pattern.cwe_id
      "CWE-328"
  """
  def weak_crypto_md5 do
    %Pattern{
      id: "ruby-weak-crypto-md5",
      name: "Weak Cryptography - MD5 Usage",
      description: "Detects usage of weak MD5 hash algorithm",
      type: :cryptographic_failure,
      severity: :medium,
      languages: ["ruby"],
      regex: [
        ~r/Digest::MD5/,
        ~r/OpenSSL::Digest(?:\.new\(['"]MD5['"]\)|::MD5)/,
        ~r/\.md5\s*\(/
      ],
      default_tier: :public,
      cwe_id: "CWE-328",
      owasp_category: "A02:2021",
      recommendation: "Use SHA-256 or SHA-384 for cryptographic hashing. For password hashing, use bcrypt.",
      test_cases: %{
        vulnerable: [
          ~S|Digest::MD5.hexdigest(password)|,
          ~S|OpenSSL::Digest.new('MD5')|,
          ~S|require 'digest'
hash = Digest::MD5.hexdigest(data)|
        ],
        safe: [
          ~S|Digest::SHA256.hexdigest(data)|,
          ~S|BCrypt::Password.create(password)|,
          ~S|OpenSSL::Digest.new('SHA256')|
        ]
      }
    }
  end
  
  
  @doc """
  SQL Injection via String Interpolation pattern.
  
  Detects SQL queries built with string interpolation.
  
  ## Examples
  
      iex> pattern = RsolvApi.Security.Patterns.Ruby.sql_injection_interpolation()
      iex> pattern.type
      :sql_injection
  """
  defdelegate sql_injection_interpolation(), to: SqlInjectionInterpolation, as: :pattern
  
  @doc """
  Command Injection pattern.
  
  Detects shell command execution with user input.
  
  ## Examples
  
      iex> pattern = RsolvApi.Security.Patterns.Ruby.command_injection()
      iex> pattern.severity
      :critical
  """
  defdelegate command_injection(), to: CommandInjection, as: :pattern
  
  @doc """
  XPath Injection pattern.
  
  Detects XPath queries with user input.
  
  ## Examples
  
      iex> pattern = RsolvApi.Security.Patterns.Ruby.xpath_injection()
      iex> pattern.type
      :xpath_injection
  """
  defdelegate xpath_injection(), to: XpathInjection, as: :pattern
  
  @doc """
  LDAP Injection pattern.
  
  Detects LDAP queries with unsanitized user input.
  
  ## Examples
  
      iex> pattern = RsolvApi.Security.Patterns.Ruby.ldap_injection()
      iex> pattern.cwe_id
      "CWE-90"
  """
  defdelegate ldap_injection(), to: LdapInjection, as: :pattern
  
  @doc """
  Weak Random Number Generation pattern.
  
  Detects use of predictable random number generators.
  
  ## Examples
  
      iex> pattern = RsolvApi.Security.Patterns.Ruby.weak_random()
      iex> pattern.severity
      :medium
  """
  defdelegate weak_random(), to: WeakRandom, as: :pattern
  
  @doc """
  Debug Mode Enabled pattern.
  
  Detects debugging code in production.
  
  ## Examples
  
      iex> pattern = RsolvApi.Security.Patterns.Ruby.debug_mode_enabled()
      iex> pattern.type
      :information_disclosure
  """
  defdelegate debug_mode_enabled(), to: DebugModeEnabled, as: :pattern
  
  @doc """
  Dangerous Eval Usage pattern.
  
  Detects eval usage with user input.
  
  ## Examples
  
      iex> pattern = RsolvApi.Security.Patterns.Ruby.eval_usage()
      iex> pattern.severity
      :critical
  """
  defdelegate eval_usage(), to: EvalUsage, as: :pattern
  
  
  
  @doc """
  Weak Password Storage pattern.
  
  Detects insecure password storage methods.
  
  ## Examples
  
      iex> pattern = RsolvApi.Security.Patterns.Ruby.weak_password_storage()
      iex> pattern.type
      :cryptographic_failure
  """
  defdelegate weak_password_storage(), to: WeakPasswordStorage, as: :pattern
  
  @doc """
  Unsafe Deserialization - Marshal pattern.
  
  Detects unsafe use of Marshal.load with user input.
  
  ## Examples
  
      iex> pattern = RsolvApi.Security.Patterns.Ruby.unsafe_deserialization_marshal()
      iex> pattern.severity
      :critical
  """
  defdelegate unsafe_deserialization_marshal(), to: UnsafeDeserializationMarshal, as: :pattern
  
  @doc """
  Unsafe YAML Loading pattern.
  
  Detects unsafe YAML deserialization vulnerabilities.
  
  ## Examples
  
      iex> pattern = RsolvApi.Security.Patterns.Ruby.unsafe_yaml()
      iex> pattern.type
      :deserialization
  """
  defdelegate unsafe_yaml(), to: UnsafeYaml, as: :pattern
  
  @doc """
  Insufficient Security Logging pattern.
  
  Detects missing security event logging that could prevent incident detection.
  
  ## Examples
  
      iex> pattern = RsolvApi.Security.Patterns.Ruby.insufficient_logging()
      iex> pattern.severity
      :medium
  """
  defdelegate insufficient_logging(), to: InsufficientLogging, as: :pattern
  
  @doc """
  SSRF via open-uri pattern.
  
  Detects Server-Side Request Forgery vulnerabilities.
  
  ## Examples
  
      iex> pattern = RsolvApi.Security.Patterns.Ruby.ssrf_open_uri()
      iex> pattern.type
      :ssrf
  """
  defdelegate ssrf_open_uri(), to: SsrfOpenUri, as: :pattern
  
  @doc """
  XSS in ERB Templates pattern.
  
  Detects cross-site scripting vulnerabilities in ERB.
  
  ## Examples
  
      iex> pattern = RsolvApi.Security.Patterns.Ruby.xss_erb_raw()
      iex> pattern.type
      :xss
  """
  defdelegate xss_erb_raw(), to: XssErbRaw, as: :pattern
  
  @doc """
  Path Traversal pattern.
  
  Detects file access with user-controlled paths.
  
  ## Examples
  
      iex> pattern = RsolvApi.Security.Patterns.Ruby.path_traversal()
      iex> pattern.severity
      :high
  """
  defdelegate path_traversal(), to: PathTraversal, as: :pattern
  
  @doc """
  Open Redirect pattern.
  
  Detects unvalidated redirect destinations.
  
  ## Examples
  
      iex> pattern = RsolvApi.Security.Patterns.Ruby.open_redirect()
      iex> pattern.type
      :open_redirect
  """
  defdelegate open_redirect(), to: OpenRedirect, as: :pattern
  
  @doc """
  Insecure Cookie Settings pattern.
  
  Detects cookies without security flags.
  
  ## Examples
  
      iex> pattern = RsolvApi.Security.Patterns.Ruby.insecure_cookie()
      iex> pattern.severity
      :medium
  """
  defdelegate insecure_cookie(), to: InsecureCookie, as: :pattern
end